src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -188,42 +189,44 @@ class BlockFetcher(
         case Left(_) => state
         case Right(validHashes) => state.withPossibleNewTopAt(validHashes.lastOption.map(_.number))
       }
-
       supervisor ! ProgressProtocol.GotNewBlock(newState.knownTop)
-
       fetchBlocks(newState)
     case MessageFromPeer(NewBlock(_, block, _), peerId) =>
+      log.debug("Received NewBlock {}", block.idTag)
       val newBlockNr = block.number
       val nextExpectedBlock = state.lastFullBlockNumber + 1
-
-      log.debug("Received NewBlock nr {}", newBlockNr)
-
       // we're on top, so we can pass block directly to importer
       if (newBlockNr == nextExpectedBlock && state.isOnTop) {
-        log.debug("Pass block directly to importer")
+        log.debug("Passing block directly to importer")
         val newState = state.withPeerForBlocks(peerId, Seq(newBlockNr)).withKnownTopAt(newBlockNr)
         state.importer ! OnTop
         state.importer ! ImportNewBlock(block, peerId)
         supervisor ! ProgressProtocol.GotNewBlock(newState.knownTop)
         context become started(newState)
         // there are some blocks waiting for import but it seems that we reached top on fetch side so we can enqueue new block for import
       } else if (newBlockNr == nextExpectedBlock && !state.isFetching && state.waitingHeaders.isEmpty) {
-        log.debug("Enqueue new block for import")
+        log.debug("Enqueueing new block for import")
         val newState = state.appendNewBlock(block, peerId)
         supervisor ! ProgressProtocol.GotNewBlock(newState.knownTop)
         context become started(newState)
-        // waiting for some bodies but we don't have this header yet - at least we can use new block header
+        // waiting for some bodies but we don't have this header yet - at least we can try to use new block header
       } else if (newBlockNr == state.nextToLastBlock && !state.isFetchingHeaders) {
```

Comment:
I think so, it may be the case where we fill the block gap between fetcher and actual blockchain without much hassle just by receiving some new block from the network

Reply:
But shouldn't the condition be `state.waitingHeaders.lastOption.map(_.number).getOrElse(state.nextToLastBlock)`, if not the append will fail

Why not keep it simple and attempt appending it all the time?

Reply:
I'm not sure I get your point - do you want to simply remove this branch or merge all remaining branches to single, where we try to append header of new block unconditionally and then see what happens?

I like the latter approach TBH.

Reply:
And on top of that it would even be quite simple to try to find suffix of headers that can be appended to queue so even response from pending request could be at least partially used. (In future/separate task of course)

Reply:
The latter as well! That would simplify quite a lot this code

>And on top of that it would even be quite simple to try to find suffix of headers that can be appended to queue so even response from pending request could be at least partially used. (In future/separate task of course)

Hmm I'd prefer adding it here but you're right that it's not as straightforward (ideally we would add some comparison between the branch being appended and the one in memory), could you create a task for it?

Reply:
IMHO, if we are still fetching things, we should be able to not use current headers from new block, i mean, other stuff should be sufficient to receive the missing info later on.

Reply:
tbh, i think that this handling is trying to do a lot at the same time, and feels like the reason behind that is because is trying to patch other parts of the process.

Reply:
I totally agree, other reason is that it was created only with following the network in mind and handling mining blocks was added at the very end as an afterthought just to enable local development.